From 31b1e3e64558f7d5e825cc353f533912312ab674 Mon Sep 17 00:00:00 2001
From: Zhihao Cheng <chengzhihao1@huawei.com>
Date: Mon, 15 May 2023 11:24:32 +0800
Subject: [PATCH 014/263] [Huawei] jbd2: Fix null ptr dereference in
 jbd2_log_do_checkpoint

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188770

--------------------------------

Following process will trigger a null ptr dereference in
jbd2_log_do_checkpoint:

          P1                P2
jbd2_journal_commit_transaction
 journal->j_checkpoint_transactions = commit_transaction
 // Add trans into journal->j_checkpoint_trans
 commit_transaction->t_state = T_COMMIT_CALLBACK;
 write_unlock(&journal->j_state_lock); // unlocking state
		     ext4_mkdir
		      start_this_handle
		       add_transaction_credits
			__jbd2_log_wait_for_space
			 jbd2_log_do_checkpoint
			  __jbd2_journal_remove_checkpoint
			    __buffer_unlink(jh)
			     transaction->t_checkpoint_list = NULL
			      if (transaction->t_state != T_FINISHED)
				return 0
			  jh2bh(transaction->t_checkpoint_list)
			  // transaction->t_checkpoint_list is NULL !

Fix it by breaking out transaction->t_checkpoint_list iteration if
transaction->t_checkpoint_list becomes NULL after invoking
__jbd2_journal_remove_checkpoint().

Fixes: a52de4f76d25 ("[Backport] jbd2: recheck chechpointing non-dirty buffer")
Signed-off-by: Zhihao Cheng <chengzhihao1@huawei.com>
---
 fs/jbd2/checkpoint.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/fs/jbd2/checkpoint.c b/fs/jbd2/checkpoint.c
index 740ee2fc726f..6edb94b77ffc 100644
--- a/fs/jbd2/checkpoint.c
+++ b/fs/jbd2/checkpoint.c
@@ -254,8 +254,12 @@ int jbd2_log_do_checkpoint(journal_t *journal)
 		}
 		if (!buffer_dirty(bh)) {
 			BUFFER_TRACE(bh, "remove from checkpoint");
-			if (__jbd2_journal_remove_checkpoint(jh))
-				/* The transaction was released; we're done */
+			/*
+			 * If the transaction was released or the checkpoint
+			 * list was empty, we're done.
+			 */
+			if (__jbd2_journal_remove_checkpoint(jh) ||
+			    !transaction->t_checkpoint_list)
 				goto out;
 		} else {
 			/*
-- 
2.17.1

